File: 2.链表/5.两两交换链表中的节点/24.两两交换链表中的节点.cpp
/*
 * @lc app=leetcode.cn id=24 lang=cpp
 *
 * [24] 两两交换链表中的节点
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

 class Solution {
    public:
        ListNode* swapPairs(ListNode* head) {
            ListNode* p = head;
            int cnt = 0;
            while (p)
            {
                cnt ++;
                p = p->next;
            }
            if (cnt == 0) return nullptr;
            if (cnt == 1) return head;
            int step = cnt / 2;
            ListNode* dummyHead = new ListNode(0);
            dummyHead->next = head->next;
            ListNode* cur = head;
            ListNode* pre = new ListNode(0);
            pre->next = cur;
            ListNode* post = cur->next;
            while (step --)
            {
                pre->next = cur->next;
                cur->next = post->next;
                post->next = cur;
                if (!step) return dummyHead->next;
                pre = cur;
                cur = pre->next;
                post = cur->next;
            }
            return dummyHead->next;
        }
    };
// @lc code=end

File: HOT100/HOT100p5/post/24.两两交换链表中的节点.py
#
# @lc app=leetcode.cn id=24 lang=python3
#
# [24] 两两交换链表中的节点
#
from mytools import *
# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        def rev(pre, cur, k):
            pre_t = None
            cur_t = cur
            for _ in range(k):
                tmp = cur_t.next
                cur_t.next = pre_t
                pre_t = cur_t
                cur_t = tmp
            pre.next = pre_t
            cur.next = cur_t
            return pre_t, cur
        cnt = 0
        cur = head
        while cur:
            cnt += 1
            cur = cur.next
        t = cnt // 2
        if t == 0: return head
        f = False
        cur = head
        pre = ListNode(next=cur)
        for _ in range(t):
            h_, t_ = rev(pre, cur, 2)
            if not f: res = h_; f = True
            pre = t_; cur = t_.next
        return res
# @lc code=end

